Step1 과제 제출입니다.

cjsrhd94 opened the pull request:
안녕하세요 성현님! 
과제 제출이 너무 늦었네요ㅠㅠ... 
많은 리뷰 부탁드립니다 :)

cjsrhd94 committed:
chore: 이전 단계 코드 이관

cjsrhd94 committed:
test: 구간 추가 관련 인수테스트 작성

cjsrhd94 committed:
feat: 노선의 맨 앞과 중간에 구간을 등록할 수 있는 기능 구현

- 노선의 맨 앞에 구간을 등록할 수 있다.
- 노선의 중간에 구간을 등록할 수 있다.
- 이미 등록되어있는 역은 노선에 등록될 수 없다.

boorownie commented on src/test/java/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -69,20 +74,79 @@ void createSection() {
+    @Test
+    void 지하철_구간_중간에_구간_생성() {
+        ExtractableResponse<Response> response
+                = RestAssuredUtil.post(SECTION_THREE, "/lines/" + LINE_SHINBUNDANG_ID + "/sections");
```
RestAssuredUtil로 요청에 대한 추상화를 하셨군요!
메서드 인자로 넘기는 값에 대한 가독성을 고려했을 때 builder 패턴을 활용해도 좋겠다는 생각이 드네요
참고만 해주세요~

boorownie commented on src/test/java/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -69,20 +74,79 @@ void createSection() {
+     * When 지하철 구간을 중간에 추가할 때 기존 구간보다 거리가 길면
+     * Then IllegalArgumentException이 발생한다.
+     */
+    @DisplayName("지하철 구간을 중간에 추가할 때 기존 구간보다 거리가 길면 추가할 수 없다.")
```
비슷한 요청에 대한 테스트가 많아 보이는데요~
비즈니스 요구사항을 검증하는 목적으로 보았을 때 인수 테스트와 단위 테스트는 어떤 차이가 있을까요?
효과가 비슷하다면 비용이 적게드는 쪽으로 구성해보면 어떨까요~?

cjsrhd94 replied:
이 부분은 이전 멘토님과도 이야기 해본 부분이었어요! 
현재 과정이 인수테스트 중심으로 구성되어있기 때문에 살짝 일부러(?) 인수테스트로 작성하였습니다. 저는 개인적으로 특정 에러가 발생하는 케이스의 경우 기존의 인수테스트 작성방식보다는 새로 작성한 단위테스트로 구체적으로 검증하는 것이 더 좋다고 생각하는데 성현님은 어떻게 생각하시나요? 아니면 인수 테스트를 조금 더 보완할 수 있는 부분이 있을까요?

boorownie replied:
비즈니스 로직에 대한 검증은 단위 테스트로 구체적으로 하는편입니다.
특히 API 요청이나 시나리오는 같은데 인자(값)만 다른 경우 더욱 단위 테스트 위주로 검증하고
시나리오가 달라지는 경우 인수 테스트로 검증하는 편입니다!

boorownie commented on src/test/java/subway/acceptance/SectionAcceptanceTest.java:
```diff
@@ -69,20 +74,79 @@ void createSection() {
-     * Then IllegalArgumentException이 발생한다.
+     * When 지하철 구간을 맨 앞에 생성하면
+     * Then 지하철 노선 조회 시 생성한 구간을 찾을 수 있다
      */
-    @DisplayName("신규 구간의 상행역과 노선의 하행종점역이 같아야 한다.")
+    @DisplayName("지하철 구간을 처음에 생성할 수 있다.")
     @Test
-    void cant_add_section_when_upStation_in_new_section_is_not_equal_line_downStation() {
+    void 지하철_노선_맨_앞에_구간_생성() {
```
테스트 로직으로는 인수 테스트에서 어떤 시나리오를 검증하는지 한번에 파악하기는 어렵지만
given, when, then에 맞춰 구성해주셔서 그런지 주석을 통해 충분히 나타내고 있군요!

boorownie commented on src/main/java/subway/section/Sections.java:
```diff
@@ -38,17 +38,79 @@ public boolean isLastSection(Station station) {
-        if (!sections.isEmpty()) {
-            if (!isLastSection(section.getUpStation())) {
-                throw new IllegalArgumentException("상행역이 하행종점역과 같지 않습니다.");
+    public void add(Section newSection) {
```
어떻게보면 이번 주차의 핵심 비즈니스 로직으로 볼 수 있겠네요.
해당 로직이 도메인 클래스 내에 있으니 테스트 작성은 쉬울 것 같은데
단위 테스트를 통해 해당 로직을 검증해보면 어떨까요?

cjsrhd94 replied:
단위 테스트 추가하였습니다!

boorownie commented on src/main/java/subway/section/Sections.java:
```diff
@@ -2,16 +2,16 @@
-    @OneToMany(mappedBy = "line", cascade = CascadeType.PERSIST, orphanRemoval = true)
+    @OneToMany(cascade = CascadeType.PERSIST, orphanRemoval = true)
+    @JoinColumn(name = "line_id")
+    @OrderColumn
     private List<Section> sections = new ArrayList<>();
```
@OrderColumn을 이용할 경우 위치가 변경될 때 마다 포지션을 의미하는 컬럼의 수정이 많이 발생합니다.
변화가 많이 발생하지 않을 수 있는 서비스의 요구사항이지만 이런점을 고려해서 그래프 형식으로 설계를 다시해보면 어떨까요?

cjsrhd94 replied:
그래프 형식으로 변경하였습니다!

boorownie commented on src/main/java/subway/section/Sections.java:
```diff
@@ -38,17 +38,79 @@ public boolean isLastSection(Station station) {
+        for (int position = 0; position < num; position++) {
+            // newSection의 upstation과 originSection의 upstation이 같을 때 구간을 중간에 추가한다.
+            Section originSection = sections.get(position);
+            if (newSection.equalUpStation(originSection.getUpStation())) {
```
케이스에 따라서 수행하는 로직이 달라지는군요!
early return을 통해 else를 피할 수 있어 좋은 것 같아요.

조금 더 욕심을 내자면 전략 패턴을 활용해서 조건(`newSection.equalUpStation(originSection.getUpStation())`) 에 따라 수행하는 로직을 분리해줄 수 도 있겠네요 :) 

반드시 반영해야 하는 건 아니니 참고만 해주세요~

boorownie commented on src/main/java/subway/section/Section.java:
```diff
@@ -58,10 +66,22 @@ public Line getLine() {
         return downStation.equals(station);
     }
 
+    public boolean isDistanceLowerOrEqual(Section section) {
+        return distance <= section.getDistance();
+    }
```
행위를 명시적으로 알려주는 메서드 네이밍 👍

boorownie reviewed (changes requested):
안녕하세요! 리뷰를 맡은 류성현입니다.
절대 늦지 않았어요! 지금부터라도 열심히 하시면 분명 큰 도움이 되실껍니다.

로직 구현을 간소하게 잘 해주셨는데 한가지 다시 확인해주시면 좋을 부분을 말씀드리자면
@OrderColumn을 이용하여 구간의 순서를 보장해주셨는데 실제 JPA를 사용할 경우 장점 보다는 단점이 큰 방법이라고 할 수 있습니다.

이 방법 대신 그래프를 이용하여 구현해보시는 것을 추천드리는데요, 이는 조금 더 유연하게 객체 설계를 할 수 있다는 점과 뒤에 있을 경로 조회에 도움이 될 수 있습니다.

나머지 코멘트는 참고만 해주시고 이어서 진행해주세요!
혹시 코멘트에 대해 추가 설명이 필요하시다면 DM으로 알려주세요~

cjsrhd94 committed:
refactor: @OrderColumn 삭제 및 역간 최단거리 정렬 기능 구현

- 기존 @OrderColumn의 성능 문제로 인해 역간 최단거리 정렬 기능을 구현하였습니다.
- 향후 여러 노선에서 사용할 것을 대비하여 Path 클래스로 분리 구현하였습니다.

cjsrhd94 committed:
test: Sections add 로직 테스트 추가

- 신규 노선의 상행역과 기존 노선의 상행역이 같은 케이스 테스트 추가
- 신규 노선의 하행역과 기존 노선의 하행역이 같을 케이스 테스트 추가
- 신규 노선이 노선 맨 앞에 추가되는 케이스 테스트 추가

cjsrhd94 committed:
test: 구간 추가시 유효성 검사 단위 테스트 작성

- 동일 지하철 구간 유효성 검사 테스트 작성
- 기존 지하철 구간 사이에 신규 구간 생성시 유효성 검사 테스트 작성

cjsrhd94 commented:
요청주신 변경사항들을 수정하면서 궁금한 점이 있었습니다.
1) JPA 엔티티에서 equals and hashcode를 구현할 때 어떻게 구현하는 것이 좋은 방법일까요?

boorownie commented on src/main/java/subway/section/Path.java:
```diff
@@ -0,0 +1,54 @@
+                graph.setEdgeWeight(graph.addEdge(section.getUpStation(), section.getDownStation()), section.getDistance());
+            }
+        }
+        // 다익스트라 알고리즘을 통한 최단거리 경로 계산
+        DijkstraShortestPath<Station, DefaultWeightedEdge> shortestPath = new DijkstraShortestPath<>(graph);
+        GraphPath<Station, DefaultWeightedEdge> path = shortestPath.getPath(start, end);
+        return path.getVertexList();
```
헉 이 부분은 3단계 요구사항인데!! ㅋㅋㅋ

boorownie commented on src/main/java/subway/section/Sections.java:
```diff
@@ -5,13 +5,10 @@
-import java.util.stream.Collectors;
 
 @Embeddable
 public class Sections {
```
그래프 형태로의 전환 👍

boorownie commented on src/test/java/subway/unit/LineTest.java:
```diff
@@ -62,4 +68,64 @@ void removeSection() {
+     *  Then 추가된 구간들을 조회할 수 있다.
+     */
+    @Test
+    void 노선_맨_앞에_구간을_추가할_수_있다() {
```
풍부한 단위 테스트 👍

boorownie commented on src/main/java/subway/section/Section.java:
```diff
@@ -58,10 +78,22 @@ public Line getLine() {
+    }
+
     @Override
     public boolean equals(Object o) {
```
질문 주신 euqals와 hashcode에 대해서는 고민이 필요할 것 같아요.
보통 식별자를 통해서 동등성을 비교할 수 있는데 그렇지 않은 케이스도 많죠
예를 들면 구간과 같은 경우에는 upstation와 downstation으로 식별하게끔 로직을 구현할 수 있으니
케이스에 따라서 식별자로 구분을 하는지 아니면 비즈니스 키로구분하는지에 따라 다를 것 같아요.
이 부분을 조금 고려해야 할 것 같네요 :)

boorownie reviewed (approved):
피드백 반영 고생하셨습니다. 
조금 까다로운 피드백이였는데 잘 수행해주셨네요 :) 
질문 주신 부분은 코멘트로 의견 남겼습니다!
이어서 2단계 진행해주세요~
